packages/access-api/src/service/upload-api-proxy.js

```diff
@@ -0,0 +1,266 @@
+ */
+function createInvocationResponder(options) {
+  /**
+   * @template {import('@ucanto/interface').Capability} Capability
+   * @param {iucanto.Invocation<Capability>} invocationIn
+   * @param {iucanto.InvocationContext} context
+   * @returns {Promise<iucanto.Result<any, { error: true }>>}
+   */
+  return async function handleInvocation(invocationIn, context) {
+    const connection = options.connections.get(invocationIn.audience.did())
+    if (!connection) {
+      throw new Error(
+        `unable to get connection for audience ${invocationIn.audience.did()}}`
+      )
+    }
+    // eslint-disable-next-line unicorn/prefer-logical-operator-over-ternary
+    const proxyInvocationIssuer = options.signer
+      ? // this results in a forwarded invocation, but the upstream will reject the signature
+        // created using options.signer unless options.signer signs w/ the same private key as the original issuer
+        // and it'd be nice to not even have to pass around `options.signer`
+        options.signer
+      : // this works, but involves lying about the issuer type (it wants a Signer but context.id is only a Verifier)
+        // @Gozala can we make it so `iucanto.InvocationOptions['issuer']` can be a Verifier and not just Signer?
+        /** @type {ed25519.Signer.Signer} */ (context.id)
+
+    const [result] = await Client.execute(
+      [
+        Client.invoke({
+          issuer: proxyInvocationIssuer,
+          capability: invocationIn.capabilities[0],
+          audience: invocationIn.audience,
+          proofs: [invocationIn],
+        }),
+      ],
+      /** @type {Client.ConnectionView<any>} */ (connection)
+    )
+    return result
+  }
+}
+
+/**
+ * @template {Record<string, any>} T
+ * @param {object} options
+ * @param {iucanto.Signer} [options.signer]
+ * @param {Pick<Map<dagUcan.DID, iucanto.ConnectionView<T>>, 'get'>} options.connections
+ */
+function createProxyStoreService(options) {
+  const handleInvocation = createInvocationResponder(options)
+  /**
+   * @type {StoreService}
+   */
+  const store = {
+    add: handleInvocation,
+    list: handleInvocation,
+    remove: handleInvocation,
+  }
+  return store
+}
+
+/**
+ * @template {Record<string, any>} T
+ * @param {object} options
+ * @param {iucanto.Signer} [options.signer]
+ * @param {Pick<Map<dagUcan.DID, iucanto.ConnectionView<T>>, 'get'>} options.connections
+ */
+function createProxyUploadService(options) {
+  const handleInvocation = createInvocationResponder(options)
+  /**
+   * @type {UploadService}
+   */
+  const store = {
+    add: handleInvocation,
+    list: handleInvocation,
+    remove: handleInvocation,
+  }
+  return store
+}
+
+/**
+ * @typedef UcantoHttpConnectionOptions
+ * @property {dagUcan.DID} audience
+ * @property {typeof globalThis.fetch} options.fetch
+ * @property {URL} options.url
+ */
+
+/**
+ * @implements {Pick<Map<dagUcan.DID, iucanto.Connection<any>>, 'get'>}
+ */
+class AudienceConnections {
```

Comment:
Nit: This whole class seems like incidental complexity, things would be a whole lot simpler with just `config` object and static function for `get` method.

P.S.: I know that's not a popular opinion, yet I like to share it in a hope to bring peers to a true side :P

Reply:
Is this like what you had in mind? https://github.com/web3-storage/w3protocol/pull/334/commits/b46dd85598c427e2a5b0e86fba2f9cb76342f336

> P.S.: I know that's not a popular opinion, yet I like to share it in a hope to bring peers to a true side :P

Eh, I don't disagree necessarily. good suggestion.

Reply:
Not exactly, sorry for not been very clear. In linked version you still have `audienceConnections` which has `get` function. When I criticize classes what I mean is:

Separate data part which is basically a state from the behavior which are all the methods

That way it does not matter how you constructed state, you can just pass it into a function and it will compute whatever. Separating data and behavior de-tangles two allowing you to store data in e.g. indexDB and then apply function after reading it from indexDB without having to built a class first. Not the case here but complex class heirerchies also often entangle initialization order.

I think there are only two valid reasons to resort to a classes with methods:

1. When you have to deal with open-ended variant space. In other sometimes you have many different state options but only one of them will be used in practice. In this scenario dispatching from one function introduces unnecessary coordination and defining interface is better option.
2. When your methods are simply a convenience sugar for the static functions. I do it all the time to allow method chaining and remove need for importing lot of dependencies. In this case they don't tangle anything they are nothing but a sugar


I would argue use of classes with methods for any other reason just introduces incidental complexity and gets between you and data nothing else.

Reply:
@Gozala ah I understand what you mean now.

> Separating data and behavior de-tangles two allowing you to store data in e.g. indexDB and then apply function after reading it from indexDB without having to built a class first.

def have run into that. That helped me understand.